node/node_newblock.go

```diff
@@ -144,13 +145,15 @@ func (node *Node) proposeShardState(block *types.Block) error {
 
 func (node *Node) proposeBeaconShardState(block *types.Block) error {
 	// TODO ek - replace this with variable epoch logic.
-	if !core.IsEpochLastBlock(block) {
+	if !shard.Schedule.IsLastBlock(block.Number().Uint64()) {
 		// We haven't reached the end of this epoch; don't propose yet.
 		return nil
 	}
-	nextEpoch := new(big.Int).Add(block.Header().Epoch(), common.Big1)
-	// TODO: add logic for EPoS
-	shardState, err := core.CalculateNewShardState(node.Blockchain(), nextEpoch)
+	// TODO Use ReadFromComputation
+	prevEpoch := new(big.Int).Sub(block.Header().Epoch(), common.Big1)
```

Comment:
Why use prevEpoch, instead of the original nextEpoch?

Reply:
The original logic confused me:

Original:
1. Take the block's header.epoch, then add one. 
https://github.com/harmony-one/harmony/blob/master/node/node_newblock.go#L151
2. Call CalculateNewShardState and decrement the epoch by 1
https://github.com/harmony-one/harmony/blob/master/core/resharding.go#L159

Calculate doesn't actually compute anything, just reads off the chain in non-epoch == 0 case. 

So actually it always ends up reading from same epoch, No?